crates/full-node/sov-db/src/historical_state.rs

```diff
@@ -1,70 +1,134 @@
 use std::fmt::Debug;
+use std::sync::Arc;
 
 use rockbound::cache::delta_reader::DeltaReader;
+use rockbound::versioned_db::{VersionedDeltaReader, VersionedSchemaBatch};
 use rockbound::{SchemaBatch, SchemaKey, SchemaValue};
 use sov_rollup_interface::common::SlotNumber;
 
 use crate::metrics::StateMaterializationMetrics;
 use crate::namespaces::{KernelNamespace, Namespace, UserNamespace};
-use crate::schema::namespace::StateValues;
+use crate::schema::namespace::NomtStateValues;
 use crate::schema::tables::StateRootHashes;
 use crate::schema::types::StateRootHashId;
-use crate::{ensure_version_is_correct, DbOptions};
+use crate::DbOptions;
 
 const STATE_ROOT_HASH_SINGLETON: StateRootHashId = StateRootHashId(0);
 
 /// A typed wrapper around the [`DeltaReader`] for reading materializing historical rollup state.
 #[derive(Debug, Clone)]
 pub struct HistoricalStateReader {
     /// The underlying [`DeltaReader`] correctly routes requests to previous snapshots and/or [`rockbound::DB`]
-    db: DeltaReader,
+    user: VersionedDeltaReader<NomtStateValues<UserNamespace>>,
+    /// The underlying [`DeltaReader`] correctly routes requests to previous snapshots and/or [`rockbound::DB`]
+    kernel: VersionedDeltaReader<NomtStateValues<KernelNamespace>>,
+    other: DeltaReader,
+
     /// The [`SlotNumber`] that will be used for the next batch of writes to the DB
     /// This [`SlotNumber`] is also used for querying data,
     /// so if this instance of [`HistoricalStateReader`] is used as read-only, it won't see newer data.
     next_version: SlotNumber,
 }
 
+/// A collection of changes to the state db. Includes versioned changes to user/kernel state, and a plain schema batch of changes to any other columns.
+#[derive(Debug, Clone, Default)]
+pub struct StateChanges {
+    pub(crate) user: Arc<VersionedSchemaBatch<NomtStateValues<UserNamespace>>>,
```

Comment:
Why do we need 3 Arcs here, instead of 1 around whole `StateChanges` ?

Reply:
I didn't fully appreciate that at the time of the previous discussion, but having attempted the change it's much messier with a single Arc than with 3